anonymizer/OpenDiffix.Service/Led.fs

```diff
@@ -43,244 +28,174 @@ let private mergeGivenAggregatorsInto (targetBucket: Bucket) aggIndexes (sourceB
+
+  victimBuckets, caches
+
+/// Main LED logic.
+let private led (aggregationContext: AggregationContext) (buckets: Bucket array) =
+  let diagnostics = LedDiagnostics()
+  let victimBuckets, caches = prepareBuckets aggregationContext buckets
+
+  let stopwatch = startStopwatch ()
 
   let anonAggregators = anonymizingAggregatorIndexes aggregationContext
+  let groupingLabelsLength = aggregationContext.GroupingLabels.Length
 
-  // Fast lookup for checking isolating column conditions.
-  // Buckets are grouped by their `nonIsolatingColumns`, meaning dictionary
-  // entries are lists of buckets with different values of `isolatingColumns`.
-  // We compare victims against buckets in that list. If there is exactly
-  // one high-count sibling, then we have found an isolating column.
-  let isolatorCache = Dictionary<Row, MutableList<Bucket * bool>>(Row.equalityComparer)
-
-  // Victim buckets are low-count and have <= 3 rows.
-  // While iterating, we also put entries in `isolatorCache`.
-  let pickVictimBucket bucket =
-    let lowCount = Utils.isLowCount lowCountIndex bucket
-    let cacheKey = nonIsolatingColumns |> Array.map (fun i -> bucket.Group.[i])
-    let siblings = isolatorCache |> Dictionary.getOrInit cacheKey (fun _ -> MutableList())
-    siblings.Add((bucket, lowCount))
-    if lowCount then diagnostics.IncrementBucketsLowCount()
-    if lowCount && bucket.RowCount <= 3 then Some(bucket, cacheKey) else None
-
-  let victimBuckets = buckets |> Array.choose pickVictimBucket
-
-  // Main victim bucket loop
-  for victimBucket, cacheKey in victimBuckets do
-    let victimGroup = victimBucket.Group
-    let siblingsPerColumn = Array.create groupingLabelsLength NoBuckets
-    let isolatorCandidates = isolatorCache.[cacheKey]
-
-    // Test for isolating columns
-    for otherBucket, otherLowCount in isolatorCandidates do
-      if not (obj.ReferenceEquals(victimBucket, otherBucket)) then
-        diagnostics.IncrementComparisons()
-
-        match findSingleNonMatchingColumn groupingLabelsLength victimGroup otherBucket.Group with
-        | Some nonMatchingColumn ->
-          siblingsPerColumn.[nonMatchingColumn] <-
-            match siblingsPerColumn.[nonMatchingColumn] with
-            | NoBuckets -> SingleBucket(otherBucket, otherLowCount)
-            | _ -> MultipleBuckets
+  /// Returns `Some(victimBucket, mergeTargets)` if victimBucket has an unknown column
+  /// and has merge targets (siblings which match isolating column criteria).
+  let chooseMergeTargets (victimBucket: Bucket, cacheKeys: CacheKey array) =
+    let mutable hasUnknownColumn = false
+    let mergeTargets = MutableList()
+
+    for colIndex in 0 .. groupingLabelsLength - 1 do
+      let cacheKey = cacheKeys.[colIndex]
+      let siblings = caches.[colIndex].[cacheKey]
+
+      if siblings.Count = 1 then
+        // No siblings for this column (Count=1 means itself)
+        hasUnknownColumn <- true
+
+      if siblings.Count = 2 then
```

Comment:
hm, how about turning this double `if` into a `match with`, where by we explicitly lay out the implications of how many siblings there were, like so:

```
| 1 -> ...
| 2 -> ...
| 3 -> //do nothing here but explain why
| _ -> failwith "some explanation why this is impossible"
```
?

Asking for this, because I am still unsure I navigate this correctly, the added prose here would probably dispel the doubts

Reply:
Yeah, looks better. Fixed.